### PR TITLE
Implement Magnetostatics in TorchPME

### PR DESCRIPTION
**Pull request by E-Rum**
This branch is linked to #54 and introduces the initial implementation of magnetostatics in the existing TorchPME library. The commit provides a basic implementation for direct potential calculations along with a simple test case.

With this implementation, it’s a good time to start discussing how to structure the current magnetostatic functionality and align it with the broader scope of the library. Notable changes to the main library include:
	1.	Dipole Potential Components:
Unlike charge-based potentials (e.g., Coulomb), dipole potentials consist of two components: a scalar and a tensor part. Currently, these components are explicitly returned separately and later combined in the calculator. While this approach works for now, we should evaluate whether it remains viable as we transition to more complex Ewald-based calculators.
	2.	Calculator Structure for Dipoles:
The dipole calculator significantly differs from charge-based calculators, potentially necessitating a new base calculator tailored for magnetostatics. For now, the implementation resides in the calculators folder, but it may need restructuring.
Key differences include:
	•	Passing dipoles instead of charges.
	•	Utilizing neighbor vectors instead of neighbor distances.
	•	Returning potentials in a vectorial form, which appears more convenient from an ML perspective.

### CODE REVIEW

File: src/torchpme/potentials/potential_dipole.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            (1 + torch.cos(r_mag * (torch.pi / self.exclusion_radius))) * 0.5,
+            0.0,
+        )
+
+    def from_dist(self, vector: torch.Tensor) -> torch.Tensor:
+        """TODO: Add docstring"""
+        r_mag = torch.norm(vector, dim=1, keepdim=True)
+        scalar_potential = 1.0 / (r_mag**3)
+        r_outer = torch.einsum(
+            "bi,bj->bij", vector, vector
```
Comment (PicoCentauri) [inline src/torchpme/potentials/potential_dipole.py:57]: Might be worth for speed later to replace the einsum by outer. The latter is usually much faster.

File: src/torchpme/calculators/calculator_dipole.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
+import torch
+from torch import profiler
+
+from ..potentials import PotentialDipole
+
+
+class CalculatorDipole(torch.nn.Module):
```
Comment (PicoCentauri) [inline src/torchpme/calculators/calculator_dipole.py:7]: We should the replace the "current" Calculator with "CalculatorMonopole" ?

File: src/torchpme/potentials/potential_dipole.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            0
+        ) - r_outer * (
+            3.0 / (r_mag**5)
+            - 3.0 * torch.erfc(torch.sqrt(alpha) * r_mag) / r_mag**5
+            - 2
+            * torch.sqrt(alpha / torch.pi)
+            * (2 * alpha + 3 / r_mag**2)
+            * torch.exp(-alpha * r_mag**2)
+            / r_mag**2
+        ).unsqueeze(-1)
```
Comment (PicoCentauri) [inline src/torchpme/potentials/potential_dipole.py:99]: Can you break this up into more readable lines. 10 lines for a return statement might be a but long.

File: src/torchpme/potentials/potential_dipole.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            )
+        alpha = 1 / (2 * self.smearing**2)
+        return 2 * torch.pi / 3 * torch.sqrt((alpha / torch.pi) ** 3)
+
+    def background_correction(self) -> torch.Tensor:
+        if self.smearing is None:
+            raise ValueError(
+                "Cannot compute background correction without specifying `smearing`."
+            )
+        return self.smearing * 0.0
```
Comment (PicoCentauri) [inline src/torchpme/potentials/potential_dipole.py:131]: Always 0?

```suggestion
        return 0.0
```

then we can even remove the value error?

Comment (E-Rum) [reply src/torchpme/potentials/potential_dipole.py:131]: Here, it’s done this way to stay consistent with the pipeline. This ensures that 0.0 is returned as a `torch.tensor` with the correct device and dtype; otherwise, we would just return a float number.

Comment (E-Rum) [reply src/torchpme/potentials/potential_dipole.py:131]: In general, I have no idea whether it should always be zero or not. One would need to perform a theoretical derivation to determine whether dipole neutrality is a crucial aspect for the dipole-Ewald calculator

File: src/torchpme/calculators/calculator_dipole.py (hunk 2 above)
Comment (E-Rum) [reply src/torchpme/calculators/calculator_dipole.py:7]: Good question! We can discuss it after we merge.

File: src/torchpme/potentials/potential_dipole.py (hunk 3 above)
Comment (E-Rum) [reply src/torchpme/potentials/potential_dipole.py:99]: Makes sense! I will do that after we finish, and we’ll make sure we get the correct physical values. Then, we should do the “linting.”

**Comment by E-Rum** [comment]
Ok, from that point on, it seems to be working. Both the direct and Ewald calculators give reasonable values for two dipoles with a fixed distance. Next, I will add documentation, improve the code structure, and add tests. Meanwhile, I would love to get some help in calculating reference values for comparisons.

@PicoCentauri and I started testing it with the espressomd P3MDipole code, but it seems to tune its parameters for each calculation. We will investigate how to prevent this or find an alternative method for testing.

**Comment by E-Rum** [comment]
TODO
	•	Add documentation.
	•	Add tests to check all dtypes, devices, shapes, and compatibility.
	•	Add a test with a large system.
	•	Refactor` _validate_parameters` to correctly handle dipole-wise inputs.
	•	Optionally: change the core pipeline so that magnetostatics become inherited subclasses rather than separate scripts.

File: src/torchpme/calculators/calculator_dipole.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        dipoles: torch.Tensor,
+        cell: torch.Tensor,
+        positions: torch.Tensor,
+        neighbor_indices: torch.Tensor,
+        neighbor_vectors: torch.Tensor,
+    ):
+        r"""TODO: Add docstring"""
+        # Temporarily pass the distance tensor to the _validate_parameters function
+        _validate_parameters(
+            charges=dipoles,
```
Comment (PicoCentauri) [inline src/torchpme/calculators/calculator_dipole.py:152]: I think we should provide another function that explicitly requires that the shape of dipoles is the same as positions (N,3). Othwerwise people might pass (N,4) vectors and we get suuper weird results.

Other option is we leave this here, just call rename `charges` in `_validate_paramateres` to `weights` and check the correct shape in each `potentials`. Maybe this is an even better option.

File: tests/test_magnetostatics.py
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:1]: maybe `test_dipoles`

File: tests/test_magnetostatics.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        dtype=torch.float64,
+    )
+    pot = calculator(
+        dipoles=system.dipoles,
+        cell=system.cell,
+        positions=system.positions,
+        neighbor_indices=system.neighbor_indices,
+        neighbor_vectors=system.neighbor_vectors,
+    )
+    result = torch.einsum("ij,ij->i", pot, system.dipoles).sum()
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:69]: Did you benchmark if this is the fastest way. I know that usually einsum is not fast and a clever stack of matmul might be much faster.

File: tests/test_magnetostatics.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    assert torch.isclose(result, expected_result, atol=1e-4), (
+        f"Expected {expected_result}, but got {result}"
+    )
+
+
+@pytest.mark.parametrize(
+    ("smearing", "sr_potential"),
+    [
+        (1e10, torch.tensor(-0.2656, dtype=torch.float64)),
+        (1e-10, torch.tensor(0.0000, dtype=torch.float64)),
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:80]: Where do these values come from. Especially the first line. Is this a regression test?

File: tests/test_magnetostatics.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    ],
+)
+def test_magnetostatics_sr(smearing, sr_potential):
+    calculator = CalculatorDipole(
+        potential=PotentialDipole(smearing=smearing, dtype=torch.float64),
+        full_neighbor_list=False,
+        lr_wavelength=1.0,
+        dtype=torch.float64,
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:88]: I would add a loop over double single precision and devices - as we do it in the monopole tests.

File: tests/test_magnetostatics.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    pot = calculator._compute_rspace(
+        dipoles=system.dipoles,
+        neighbor_indices=system.neighbor_indices,
+        neighbor_vectors=system.neighbor_vectors,
+    )
+    result = torch.einsum("ij,ij->i", pot, system.dipoles).sum()
+    expected_result = sr_potential
+    assert torch.isclose(result, expected_result, atol=1e-4), (
+        f"Expected {expected_result}, but got {result}"
+    )
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:99]: You don't have to put an error message here, just use `torch.testing.assert_close`

Comment (PicoCentauri) [inline tests/test_magnetostatics.py:99]: You don't have to put an error message here, just use `torch.testing.assert_close`

File: tests/test_magnetostatics.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+
+def test_magnetostatic_ewald():
+    alpha = 1.0
+    smearing = (1 / (2 * alpha**2)) ** 0.5
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:104]: Why this function. Where does it come from?

File: tests/test_magnetostatics.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    calculator = CalculatorDipole(
+        potential=PotentialDipole(smearing=smearing, dtype=torch.float64),
+        full_neighbor_list=False,
+        lr_wavelength=0.1,
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:108]: Can you give an estimate why you chose this value?

File: tests/test_magnetostatics.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+cutoffs = [3.9986718930, 4.0000000000, 4.7363281250]
+alphas = [0.8819831493, 0.8956299559, 0.7215211182]
+energies = [frame.get_potential_energy() for frame in frames]
+forces = [frame.get_forces() for frame in frames]
+
+
+@pytest.mark.parametrize(
+    ("frame", "cutoff", "alpha", "energy", "force"),
+    zip(frames, cutoffs, alphas, energies, forces),
+)
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:137]: pytest is clever - no zip needed. See `test_random_structure`. I would also call this function the same.

File: tests/test_magnetostatics.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+    expected_result = torch.tensor(energy, dtype=torch.float64)
+    assert torch.isclose(result, expected_result, atol=1e-4), (
+        f"Expected {expected_result}, but got {result}"
+    )
+
+    forces = -torch.autograd.grad(result, positions)[0]
+    expected_forces = torch.tensor(force, dtype=torch.float64)
+    assert torch.allclose(forces, expected_forces, atol=1e-4), (
+        f"Expected {expected_forces}, but got {forces}"
+    )
```
Comment (PicoCentauri) [inline tests/test_magnetostatics.py:183]: Please also test that the gradient of the cell is not nan. We have no reference values (yet) but at least with this we can test that the backprop runs.

We can use lammps to compute the stress.

**Review by PicoCentauri — COMMENTED** [review]
Nice. You should also add the calculator and potential to the toplevel import __init__ file.

I checked the tests carefully. They are still very rough. I would recommend to copy the structure/names and logic from the monopoles. The backbone of the library are the good tests that allows us to refactor the stuff heavily and we should keep this for the dipoles as well.

Also, I am missing tests for the potential itself.

File: src/torchpme/__init__.py [hunk 14]
```diff
@@ -2,12 +2,19 @@
 
 from . import calculators, lib, potentials, prefactors, tuning  # noqa
 from ._version import __version__, __version_tuple__  # noqa
-from .calculators import Calculator, EwaldCalculator, P3MCalculator, PMECalculator
+from .calculators import (
+    Calculator,
+    CalculatorDipole,
```
Comment (PicoCentauri) [inline src/torchpme/__init__.py:7]: Should we maybe name it `DipoleEwaldCalculator`, to be consistent with the monopole classes.

File: src/torchpme/calculators/calculator_dipole.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+import torch
+from torch import profiler
+
+from .._utils import _validate_parameters
+from ..lib import generate_kvectors_for_ewald
+from ..potentials import PotentialDipole
+
+
+class CalculatorDipole(torch.nn.Module):
+    """TODO: Add docstring"""
```
Comment (PicoCentauri) [inline src/torchpme/calculators/calculator_dipole.py:12]: You mind adding docstrings here?

I can prepare and push the rst entries to your branch.

**Review by PicoCentauri — COMMENTED** [review]
Very nice. I think the only thing which is left are some cosmetical changes.

1. There are a couple of doc strings to be done
2. I had rename of test file suggested in the past
3. Maybe we should also write a simple example using the dipole test frames. maybe plotting the vector field or something... But for now it can super simple.

**Comment by PicoCentauri** [comment]
I think we also need the metatensor interface and tests that are similar to what we do in 

https://github.com/lab-cosmo/torch-pme/blob/17633bb7a34294e0b1ef598e0fe00f40e7a3f34b/tests/calculators/test_workflow.py

File: src/torchpme/__init__.py (hunk 14 above)
Comment (E-Rum) [reply src/torchpme/__init__.py:7]: Agree, makes more sense, but let's leave it as it is for now, since we are planning to refactor the classes into SR and LR parts soon anyway.

**Review by PicoCentauri — APPROVED** [review]
Very nice work! Thanks for the addition.

Merge when CI and you are happy.
